### PR TITLE
CI: add action to tag PRs for release notes

### PR DESCRIPTION
**Pull request by gmuloc**
## Change Summary

This PR adds the following:
* A custom github action to be able to label the PRs with a label that looks like: `rn: <type>[(<scope>)][!]?` where `<type>` and `<scope>` are understood as per conventional commit specification and `!` means Breaking Change. The action assumes one such label only and is able to remove a preexisting label when required. When a label does not exist it can be auto_created if the action is told so via input parameters.
* A workflow that run the action described above when PRs are merged OR when the title is modified post merge. It calls the action with the input parameter to auto create the labels when needed.
* A `.github/release.yml` file that is used as an instruction manual by Github when generating release notes for a release
* A python script `generate_release.py` which can be used to generate the previous file given the amount of potential labels that the action can generate based on the supported list of `<types>` and `<scopes>`.

The goal behind this is to generate proper release notes based on the PR titles.

## Component(s) name

`ci`

## How to test

Need to merge and start seeing the labels.

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: .github/generate_release.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    "how-to",
+    "actions",
+    "molecule",
+    "ansible",
+    "github",
+    "requirements",
+    "",  # empty scope
+]
+
+CATEGORIES = {
```
Comment (ClausHolbechArista) [inline .github/generate_release.py:26]: We should add `CI`, `Test` to a list of ignore labels

File: .github/release.yml [hunk 2]
```diff
@@ -0,0 +1,696 @@
+changelog:
+  categories:
+    - title: "`eos_config_deploy_cvp` Breaking Changes \U0001F6E0"
```
Comment (ClausHolbechArista) [inline .github/release.yml]: I believe these are the sections we want (no backticks):
- Breaking Changes
  - match anything with *(*)! (they can also be included in their regular category if that is simpler)
- Change to Requirements
  - match *(requirements)
- Fixed issues in eos_cli_config_gen
  - match Fix(eos_cli_config_gen)
- Fixed issues in eos_designs
  - match Fix(eos_designs)
- Other Fixed issues
  - match Fix(*) except the two above
- Documentation
  - match Doc(*)
- New features and enhancements in eos_cli_config_gen
  - match Feat(eos_cli_config_gen)
- New features and enhancements in eos_designs
  - match Feat(eos_designs)
- Other new features and enhancements
  - match Feat(*) except the two above 
- Other changes
  - Match anything not matched above - except the ignore list

File: .github/generate_release.py (hunk 1 above)
Comment (gmuloc) [reply .github/generate_release.py:26]: Addressed in 3d91084

File: .github/release.yml (hunk 2 above)
Comment (gmuloc) [reply .github/release.yml]: Addressed in 3d91084

**Comment by gmuloc** [comment]
As discussed updating types and scope as:

```
    types: |
            Feat
            Fix
            Cut
            Doc
            CI
            Bump
            Test
            Refactor
            Revert
    # Configure which scopes are allowed.
    scopes: |
            build_output_folders
            cvp_configlet_upload
            dhcp_provisioner
            eos_cli_config_gen
            eos_config_deploy_cvp
            eos_config_deploy_eapi
            eos_designs
            eos_snapshot
            eos_validate_state
            plugins
            requirements
```

**Review by ClausHolbechArista — APPROVED** [review]
LGTM

**Comment by ClausHolbechArista** [comment]
@gmuloc I think we should remove the types and scopes from the contribution document as well, so it is synced.

File: .github/generate_release.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python
+"""
+generate_release.py
+"""
```
Comment (tgodaA) [inline .github/generate_release.py]: Generates the release.yml file right?

File: .github/generate_release.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    "dhcp_provisioner",
+    "eos_cli_config_gen",
+    "eos_config_deploy_cvp",
+    "eos_config_deploy_eapi",
+    "eos_designs",
+    "eos_snapshot",
+    "eos_validate_state",
+    "plugins",
+    "requirements",
+    "",  # empty scope
```
Comment (tgodaA) [inline .github/generate_release.py]:
```suggestion
    "requirements"
```

File: .github/generate_release.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    # First add Breaking Changes
+    breaking_labels = []
+    for scope in SCOPES:
+        for cc_type, rn_title in CATEGORIES.items():
+            # This assumes that Doc, Test and CI cannot be breaking
+            if cc_type in ["Feat", "Fix", "Cut", "Revert", "Refactor", "Bump"]:
+                if scope != "":
+                    breaking_labels.append(f"rn: {cc_type}({scope})!")
+                else:
+                    breaking_labels.append(f"rn: {cc_type}!")
```
Comment (tgodaA) [inline .github/generate_release.py]:
```suggestion
    breaking_label_categories = ["Feat", "Fix", "Cut", "Revert", "Refactor", "Bump"]
    # This assumes that Doc, Test and CI cannot be breaking
    breaking_labels = [f"rn: {cc_type}({scope})!" for cc_type in breaking_label_categories for scope in SCOPES]
    breaking_labels.extend([f"rn: {cc_type}!" for cc_type in breaking_label_categories])
```

File: .github/generate_release.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+            "labels": "rn: Fix(eos_designs)",
+        }
+    )
+    # Add other fixes
+    other_fixes_labels = []
+    for scope in SCOPES:
+        if scope not in ["eos_cli_config_gen", "eos_designs", ""]:
+            other_fixes_labels.append(f"rn: Fix({scope})")
+        elif scope == "":
+            other_fixes_labels.append("rn: Fix")
```
Comment (tgodaA) [inline .github/generate_release.py]:
```suggestion
    other_fixes_labels = [f"rn: Fix({scope})" for scope in SCOPES if scope not in ["eos_cli_config_gen", "eos_designs"]]
    other_fixes_labels.append("rn: Fix")
```

File: .github/generate_release.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    exclude_list = []
+    categories_list = []
+    # First add exclude labels
+    for scope in SCOPES:
+        if scope != "":
+            exclude_list.append(f"rn: Test({scope})")
+            exclude_list.append(f"rn: CI({scope})")
+        else:
+            exclude_list.append("rn: Test")
+            exclude_list.append("rn: CI")
```
Comment (tgodaA) [inline .github/generate_release.py]:
```suggestion
    exclude_list = []
    categories_list = []
    # First add exclude labels
    for scope in SCOPES:
        exclude_list.append(f"rn: Test({scope})")
        exclude_list.append(f"rn: CI({scope})")
    exclude_list.extend(["rn: Test", "rn: CI"])
```

File: .github/generate_release.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+            "labels": other_fixes_labels,
+        }
+    )
+    # Add Documentation
+    doc_labels = []
+    for scope in SCOPES:
+        if scope != "":
+            doc_labels.append(f"rn: Doc({scope})")
+        elif scope == "":
+            doc_labels.append("rn: Doc")
```
Comment (tgodaA) [inline .github/generate_release.py]:
```suggestion
    doc_labels = [f"rn: Doc({scope})" for scope in SCOPES]
    doc_labels.append("rn: Doc")
```

File: .github/generate_release.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+            "labels": "rn: Feat(eos_designs)",
+        }
+    )
+    # Add other new features
+    other_feat_labels = []
+    for scope in SCOPES:
+        if scope not in ["eos_cli_config_gen", "eos_designs", ""]:
+            other_feat_labels.append(f"rn: Feat({scope})")
+        elif scope == "":
+            other_feat_labels.append("rn: Feat")
```
Comment (tgodaA) [inline .github/generate_release.py]:
```suggestion
    other_feat_labels = [f"rn: Feat({scope})" for scope in SCOPES if scope not in ["eos_cli_config_gen", "eos_designs"]]
    other_feat_labels.append("rn: Feat")
```

File: .github/generate_release.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+
+    # Add the catch all
+    categories_list.append(
+        {
+            "title": "Other Changes",
+            "labels": ["*"],
+        }
+    )
+    with open(r"release.yml", "w", encoding="utf-8") as release_file:
+        yaml.dump(
```
Comment (tgodaA) [inline .github/generate_release.py:158]: Use yaml.safe_dump because dump doesn't dump properly sometimes. Not sure how it behaves here.

**Review by tgodaA — APPROVED** [review]
LGTM, approving now but I left few optional suggestions.

File: .github/generate_release.py (hunk 4 above)
Comment (carlbuchmann) [reply .github/generate_release.py]: There are cases where we don't have a scope, so we need to keep `"", # empty scope`. 
For Example: `Fix: Remove ansible-test enforced GPL3 header from modules`

Comment (tgodaA) [reply .github/generate_release.py]: Yeah, for that I handled everywhere by adding the tags manually.

Comment (tgodaA) [reply .github/generate_release.py]: For example like this: breaking_labels.extend([f"rn: {cc_type}!" for cc_type in breaking_label_categories])

**Review by carlbuchmann — APPROVED** [review]
LGTM

- Let's start testing here, and once completed, let's move this action to a dedicated repo similar [action-molecule-avd](action-molecule-avdhttps://github.com/arista-netdevops-community/action-molecule-avd) - This will allow us to use this with other AVD ecosystem repos i.e ansible-cvp.

File: .github/generate_release.py (hunk 3 above)
Comment (gmuloc) [reply .github/generate_release.py]: Addressed in cb9b7d0

File: .github/generate_release.py (hunk 4 above)
Comment (gmuloc) [reply .github/generate_release.py]: Addressed in cb9b7d0

File: .github/generate_release.py (hunk 10 above)
Comment (gmuloc) [reply .github/generate_release.py:158]: Addressed in cb9b7d0 - I am using a custom dumper to make yamllint happy but I am now deriving it from the SafeDumper.

**Pull request merged**